php-classes/Slate/CBL/Tasks/StudentTask.php

```diff
@@ -0,0 +1,162 @@
+        'Task' => [
+            'type' => 'one-one',
+            'local' => 'TaskID',
+            'class' => Task::class
+        ],
+        'Student' => [
+            'type' => 'one-one',
+            'local' => 'StudentID',
+            'class' => Person::class
+        ],
+        'Section' => [
+            'type' => 'one-one',
+            'local' => 'CourseSectionID',
+            'class' => Section::class
+        ],
+        'Comments' => [
+            'type' => 'context-children',
+            'class' => Comment::class
+        ],
+        'SkillRatings' => [
+            'type' => 'one-many',
+            'class' => StudentTaskSkillRating::class,
+            'foreign' => 'StudentTaskID',
+            'local' => 'ID'
+        ],
+        'Attachments' => [
+            'type' => 'context-children',
+            'class' => Attachments\AbstractTaskAttachment::class
+        ]
+    ];    
+    
+    public static $dynamicFields = array(
+        'Task',
+        'Student',
+        'Section',
+        'SkillRatings',
+        'Comments',
+        'Attachments',
+        'StudentName' => [
+            'getter' => 'getStudentName'
+        ],
+        'TaskSkills' => [
+            'getter' => 'getTaskSkills'
+        ]
+    );    
+
+    public static $indexes = [
+        'StudentTask' => [
+            'fields' => ['TaskID', 'StudentID']
+        ]
+    ];
+
+    public static $validators = [
+        'Task' => [
+            'validator' => 'require-relationship'
+        ],
+
+        'Student' => [
+            'validator' => 'require-relationship'
+        ]
+    ];
+    
+    public static $searchConditions = [
+        'Task' => [
+            'qualifiers' => ['task', 'task_id', 'taskid'],
+            'points' => 2,
+            'sql' => 'TaskID = %u'
+        ],
+        'Student' => [
+            'qualifiers' => ['student', 'student_id', 'studentid'],
+            'points' => 2,
+            'sql' => 'StudentID = %u'
+        ]
+    ];
+    
+    public function getStudentName()
+    {
+        return $this->Student->FullName;
+    }
+    
+    public function getTaskSkills()
+    {
+        $taskSkills = [];
+        foreach ($this->Task->Skills as $skill) {
+            $studentCompetency = $skill->Competency ? StudentCompetency::getByWhere(['StudentID' => $this->StudentID, 'CompetencyID' => $skill->Competency->ID]) : null;
+            $skillRating = StudentTaskSkillRating::getByWhere(['StudentTaskID' => $this->ID, 'SkillID' => $skill->ID]);
+            $taskSkills[] = array_merge($skill->getData(), [
+                'CompetencyLevel' => $studentCompetency ? $studentCompetency->Level : null,
+                'CompetencyCode' => $skill->Competency ? $skill->Competency->Code : null,
+                'CompetencyDescriptor' => $skill->Competency ? $skill->Competency->Descriptor : null,
+                'Rating' => $skillRating ? $skillRating->Score : null
+            ]);
+        }
+        
+        return $taskSkills;
```

Comment:
@nbey if you're putting together a deep custom data structure that's not just a calculation based on data within each record or a complex relationship, it's probably best to make a custom UI-centric data endpoint than a dynamic field